File: libs/python/mail_action_usps/rules.py
#!/usr/bin/env python3
"""
Importance rule engine for USPS mail classification.

Rules are loaded from a versioned JSON file stored in the mail agent workspace
(~/.openclaw/agents/mail/workspace/usps-mail/rules.json). The plugin ships with no
built-in rules — all rules are personal and user-managed.

Rule format:
{
  "version": "1.2",
  "rules": [
    {
      "_comment": "Former residents: Conway → low",
      "addressee_contains": "conway",
      "importance": "low"
    }
  ]
}

Supported conditions (all case-insensitive):
  <field>_contains, <field>_not_contains,
  <field>_equals, <field>_not_equals

Fields: addressee, sender, description, mail_class, address_method
"""

import json
import os

from .paths import get_rules_file

BADGE_LABELS = {
    "urgent": "🚨 Urgent",
    "high": "⚠️ Important",
    "medium": "📬 Medium",
    "low": "📭 Low",
    "junk": "🗑️ Junk",
    "ad": "📢 Ad",
    "unknown": "❓ Unknown",
}


def load_rules(rules_path: str = None, workspace_agent: str = None) -> tuple:
    """Load rules from JSON. Returns (rules_list, version_str)."""
    if not rules_path and not workspace_agent:
        raise ValueError("workspace_agent is required when rules_path is not provided")
    path = rules_path or str(get_rules_file(workspace_agent))
    if not os.path.exists(path):
        return [], "0"
    with open(path) as f:
        data = json.load(f)
    if isinstance(data, dict):
        return data.get("rules", []), str(data.get("version", "0"))
    return data, "0"


def save_rules(rules: list, version: str, rules_path: str = None, workspace_agent: str = None):
    """Atomic write of rules to disk."""
    path = rules_path or str(get_rules_file(workspace_agent))
    os.makedirs(os.path.dirname(path), exist_ok=True)
    tmp = path + ".tmp"
    with open(tmp, "w") as f:
        json.dump({"version": version, "rules": rules}, f, indent=2)
    os.replace(tmp, path)


def apply_rules(info: dict, rules: list = None) -> dict:
    """Apply importance override rules to a mailpiece info dict.

    First matching rule wins. Returns a (possibly modified) copy of info.
    """
    if rules is None:
        rules, _ = load_rules()
    if not rules:
        return info

    addressee = info.get("addressee", "").lower()
    sender = info.get("sender", "").lower()
    description = info.get("description", "").lower()
    mail_class = info.get("mail_class", "").lower()
    address_method = info.get("address_method", "").lower()

    fields = {
        "addressee": addressee,
        "sender": sender,
        "description": description,
        "mail_class": mail_class,
        "address_method": address_method,
    }

    for rule in rules:
        match = True
        for key, val in rule.items():
            if key in ("_comment", "importance"):
                continue

            if key.endswith("_not_contains"):
                field_name = key[: -len("_not_contains")]
                if val.lower() in fields.get(field_name, ""):
                    match = False
            elif key.endswith("_contains"):
                field_name = key[: -len("_contains")]
                if val.lower() not in fields.get(field_name, ""):
                    match = False
            elif key.endswith("_not_equals"):
                field_name = key[: -len("_not_equals")]
                if val.lower() == fields.get(field_name, ""):
                    match = False
            elif key.endswith("_equals"):
                field_name = key[: -len("_equals")]
                if val.lower() != fields.get(field_name, ""):
                    match = False

        if match:
            info = dict(info)
            info["importance"] = rule["importance"]
            return info

    return info


def add_rule(conditions: dict, importance: str, comment: str = "", workspace_agent: str = None) -> dict:
    """Add a new rule and bump version. Returns updated rule summary."""
    if not workspace_agent:
        raise ValueError("workspace_agent is required")
    rules, version = load_rules(workspace_agent=workspace_agent)
    new_rule = dict(conditions)
    new_rule["importance"] = importance
    if comment:
        new_rule["_comment"] = comment

    rules.append(new_rule)

    # Bump version
    try:
        major, minor = version.split(".")
        new_version = f"{major}.{int(minor) + 1}"
    except ValueError:
        new_version = f"{version}.1" if version != "0" else "1.0"

    save_rules(rules, new_version, workspace_agent=workspace_agent)
    return {"action": "added", "rule_index": len(rules) - 1, "version": new_version}


def remove_rule(index: int = None, comment_match: str = None, workspace_agent: str = None) -> dict:
    """Remove a rule by index or comment substring. Returns result."""
    if not workspace_agent:
        raise ValueError("workspace_agent is required")
    rules, version = load_rules(workspace_agent=workspace_agent)

    removed = None
    if index is not None and 0 <= index < len(rules):
        removed = rules.pop(index)
    elif comment_match:
        for i, rule in enumerate(rules):
            if comment_match.lower() in rule.get("_comment", "").lower():
                removed = rules.pop(i)
                break

    if removed is None:
        return {"action": "not_found"}

    try:
        major, minor = version.split(".")
        new_version = f"{major}.{int(minor) + 1}"
    except ValueError:
        new_version = f"{version}.1"

    save_rules(rules, new_version, workspace_agent=workspace_agent)
    return {"action": "removed", "rule": removed, "version": new_version}


def test_rule(info: dict, workspace_agent: str = None) -> dict:
    """Test which rule matches a given mailpiece. Returns match details."""
    if not workspace_agent:
        raise ValueError("workspace_agent is required")
    rules, version = load_rules(workspace_agent=workspace_agent)
    result = apply_rules(info, rules)
    original_imp = info.get("importance", "unknown")
    new_imp = result.get("importance", original_imp)

    return {
        "original_importance": original_imp,
        "final_importance": new_imp,
        "rule_matched": original_imp != new_imp,
        "rules_version": version,
    }


def list_rules(workspace_agent: str = None) -> dict:
    """List all rules with version."""
    if not workspace_agent:
        raise ValueError("workspace_agent is required")
    rules, version = load_rules(workspace_agent=workspace_agent)
    summary = []
    for i, rule in enumerate(rules):
        conditions = {k: v for k, v in rule.items() if k not in ("_comment", "importance")}
        summary.append({
            "index": i,
            "comment": rule.get("_comment", ""),
            "importance": rule.get("importance", "?"),
            "conditions": conditions,
        })
    return {"version": version, "count": len(rules), "rules": summary}

File: libs/ts/mail_action_usps/src/rules.ts
/**
 * Importance rule engine for USPS mail classification.
 *
 * Rules are loaded from a versioned JSON file stored in the mail agent workspace.
 * The plugin ships with no built-in rules — all rules are personal and user-managed.
 *
 * Supported conditions (all case-insensitive):
 *   <field>_contains, <field>_not_contains,
 *   <field>_equals, <field>_not_equals
 *
 * Fields: addressee, sender, description, mail_class, address_method
 */

import {
  existsSync,
  mkdirSync,
  readFileSync,
  renameSync,
  writeFileSync,
} from "node:fs";
import { dirname } from "node:path";

import { getRulesFile } from "./paths.js";

export const BADGE_LABELS: Record<string, string> = {
  urgent: "🚨 Urgent",
  high: "⚠️ Important",
  medium: "📬 Medium",
  low: "📭 Low",
  junk: "🗑️ Junk",
  ad: "📢 Ad",
  unknown: "❓ Unknown",
};

export interface Rule {
  importance: string;
  _comment?: string;
  [key: string]: string | undefined;
}

/**
 * Load rules from JSON. Returns [rules_list, version_str].
 */
export function loadRules(options?: {
  rulesPath?: string;
  workspaceAgent?: string;
}): [Rule[], string] {
  const opts = options ?? {};
  if (!opts.rulesPath && !opts.workspaceAgent) {
    throw new Error(
      "workspace_agent is required when rules_path is not provided",
    );
  }
  const path = opts.rulesPath ?? getRulesFile(opts.workspaceAgent);
  if (!existsSync(path)) return [[], "0"];

  const data = JSON.parse(readFileSync(path, "utf-8"));
  if (typeof data === "object" && !Array.isArray(data)) {
    return [data.rules ?? [], String(data.version ?? "0")];
  }
  return [data as Rule[], "0"];
}

/**
 * Atomic write of rules to disk.
 */
export function saveRules(
  rules: Rule[],
  version: string,
  options?: { rulesPath?: string; workspaceAgent?: string },
): void {
  const opts = options ?? {};
  const path = opts.rulesPath ?? getRulesFile(opts.workspaceAgent);
  mkdirSync(dirname(path), { recursive: true });
  const tmp = path + ".tmp";
  writeFileSync(tmp, JSON.stringify({ version, rules }, null, 2));
  renameSync(tmp, path);
}

/**
 * Apply importance override rules to a mailpiece info dict.
 * First matching rule wins. Returns a (possibly modified) copy of info.
 */
export function applyRules(
  info: Record<string, unknown>,
  rules?: Rule[],
): Record<string, unknown> {
  if (rules === undefined) {
    const [loaded] = loadRules();
    rules = loaded;
  }
  if (!rules || rules.length === 0) return info;

  const fields: Record<string, string> = {
    addressee: ((info.addressee as string) ?? "").toLowerCase(),
    sender: ((info.sender as string) ?? "").toLowerCase(),
    description: ((info.description as string) ?? "").toLowerCase(),
    mail_class: ((info.mail_class as string) ?? "").toLowerCase(),
    address_method: ((info.address_method as string) ?? "").toLowerCase(),
  };

  for (const rule of rules) {
    let match = true;
    for (const [key, val] of Object.entries(rule)) {
      if (key === "_comment" || key === "importance") continue;
      if (val === undefined) continue;

      const lowerVal = val.toLowerCase();

      if (key.endsWith("_not_contains")) {
        const fieldName = key.slice(0, -"_not_contains".length);
        if ((fields[fieldName] ?? "").includes(lowerVal)) {
          match = false;
        }
      } else if (key.endsWith("_contains")) {
        const fieldName = key.slice(0, -"_contains".length);
        if (!(fields[fieldName] ?? "").includes(lowerVal)) {
          match = false;
        }
      } else if (key.endsWith("_not_equals")) {
        const fieldName = key.slice(0, -"_not_equals".length);
        if (lowerVal === (fields[fieldName] ?? "")) {
          match = false;
        }
      } else if (key.endsWith("_equals")) {
        const fieldName = key.slice(0, -"_equals".length);
        if (lowerVal !== (fields[fieldName] ?? "")) {
          match = false;
        }
      }
    }

    if (match) {
      const copy = { ...info };
      copy.importance = rule.importance;
      return copy;
    }
  }

  return info;
}

/**
 * Add a new rule and bump version. Returns updated rule summary.
 */
export function addRule(
  conditions: Record<string, string>,
  importance: string,
  options?: { comment?: string; workspaceAgent?: string },
): { action: string; rule_index: number; version: string } {
  const opts = options ?? {};
  if (!opts.workspaceAgent) {
    throw new Error("workspace_agent is required");
  }
  const [rules, version] = loadRules({
    workspaceAgent: opts.workspaceAgent,
  });
  const newRule: Rule = { ...conditions, importance };
  if (opts.comment) {
    newRule._comment = opts.comment;
  }

  rules.push(newRule);

  let newVersion: string;
  const parts = version.split(".");
  if (parts.length === 2) {
    newVersion = `${parts[0]}.${parseInt(parts[1], 10) + 1}`;
  } else {
    newVersion = version !== "0" ? `${version}.1` : "1.0";
  }

  saveRules(rules, newVersion, { workspaceAgent: opts.workspaceAgent });
  return { action: "added", rule_index: rules.length - 1, version: newVersion };
}

/**
 * Remove a rule by index or comment substring. Returns result.
 */
export function removeRule(options: {
  index?: number;
  commentMatch?: string;
  workspaceAgent?: string;
}): { action: string; rule?: Rule; version?: string } {
  if (!options.workspaceAgent) {
    throw new Error("workspace_agent is required");
  }
  const [rules, version] = loadRules({
    workspaceAgent: options.workspaceAgent,
  });

  let removed: Rule | undefined;
  if (
    options.index !== undefined &&
    options.index >= 0 &&
    options.index < rules.length
  ) {
    removed = rules.splice(options.index, 1)[0];
  } else if (options.commentMatch) {
    const lowerMatch = options.commentMatch.toLowerCase();
    for (let i = 0; i < rules.length; i++) {
      if ((rules[i]._comment ?? "").toLowerCase().includes(lowerMatch)) {
        removed = rules.splice(i, 1)[0];
        break;
      }
    }
  }

  if (!removed) return { action: "not_found" };

  let newVersion: string;
  const parts = version.split(".");
  if (parts.length === 2) {
    newVersion = `${parts[0]}.${parseInt(parts[1], 10) + 1}`;
  } else {
    newVersion = `${version}.1`;
  }

  saveRules(rules, newVersion, { workspaceAgent: options.workspaceAgent });
  return { action: "removed", rule: removed, version: newVersion };
}

/**
 * Test which rule matches a given mailpiece. Returns match details.
 */
export function testRule(
  info: Record<string, unknown>,
  workspaceAgent?: string,
): {
  original_importance: string;
  final_importance: string;
  rule_matched: boolean;
  rules_version: string;
} {
  if (!workspaceAgent) {
    throw new Error("workspace_agent is required");
  }
  const [rules, version] = loadRules({ workspaceAgent });
  const result = applyRules(info, rules);
  const originalImp = (info.importance as string) ?? "unknown";
  const newImp = (result.importance as string) ?? originalImp;

  return {
    original_importance: originalImp,
    final_importance: newImp,
    rule_matched: originalImp !== newImp,
    rules_version: version,
  };
}

/**
 * List all rules with version.
 */
export function listRules(workspaceAgent?: string): {
  version: string;
  count: number;
  rules: Array<{
    index: number;
    comment: string;
    importance: string;
    conditions: Record<string, string>;
  }>;
} {
  if (!workspaceAgent) {
    throw new Error("workspace_agent is required");
  }
  const [rules, version] = loadRules({ workspaceAgent });
  const summary = rules.map((rule, i) => {
    const conditions: Record<string, string> = {};
    for (const [k, v] of Object.entries(rule)) {
      if (k !== "_comment" && k !== "importance" && v !== undefined) {
        conditions[k] = v;
      }
    }
    return {
      index: i,
      comment: rule._comment ?? "",
      importance: rule.importance ?? "?",
      conditions,
    };
  });
  return { version, count: rules.length, rules: summary };
}
